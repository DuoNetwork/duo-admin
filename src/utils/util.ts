import moment from 'moment';
import * as CST from '../common/constants';
import { IOption } from '../common/types';
// import { Options } from 'aws-cli-js';

class Util {
	public logLevel: string = CST.LOG_INFO;

	private log(text: any, level?: string): void {
		if (level && CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	public composeQuery(paras: { [key: string]: any }): string {
		let query: string = '?';
		for (const key in paras) query += key + '=' + paras[key] + '&';
		query = query.slice(0, -1);
		return query;
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			forceREST: argv.includes('rest'),
			live: process.argv.includes('live'),
			dbLive: process.argv.includes('dbLive'),
			server: process.argv.includes('server'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			pair: '',
			contractType: 'Beethoven',
			tenor: 'Perpetual',
			assets: [''],
			sources: [''],
			exSources: [''],
			gasPrice: 5e9,
			gasLimit: 200000,
			source: '',
			event: '',
			provider: '',
			period: 1,
			base: 'USD',
			quote: 'ETH'
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'exSources':
					option.exSources = args[1].split(',');
					break;
				case 'sources':
					option.sources = args[1].split(',');
					break;
				case 'pair':
					option.pair = args[1].replace('_', '|') || option.pair;
					break;
				case 'assets':
					option.assets = args[1].split(',');
					break;
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'period':
					option.period = Number(args[1]);
					break;
				case 'base':
					option.base = args[1] || option.base;
					break;
				case 'quote':
					option.quote = args[1] || option.quote;
					break;
				case 'contractType':
					option.contractType = args[1] || option.contractType;
					break;
				case 'tenor':
					option.tenor = args[1] || option.tenor;
					break;
				default:
					break;
			}
		}

		return option;
	}

	public getPeriodStartTimestamp(timestamp: number, period: number = 1) {
		// const now = this.getUTCNowTimestamp();
		return Math.floor(timestamp / 60000 / period - 1) * 60000 * period;
	}

	public timestampToString(ts: number) {
		//just for debugging purpose
		return moment.utc(ts).format('YYYY-MM-DDTHH:mm:ss');
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		console.log('dynamo', option.dynamo);
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case CST.TRADES:
				type = 'TRADE';
				source = option.source ? '_' + option.source.toUpperCase() : '';
				break;
			case CST.TRIGGER:
			case CST.FETCH_EVENTS:
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case CST.COMMIT:
				type = 'FEED';
				break;
			case CST.CLEAN_DB:
				type = 'CLEANDB';
				break;
			case CST.DB_PRICES:
				type = 'PRICE';
				if (option.period === 1) source = '_MINUTELY';
				else if (option.period === 60) source = '_HOURLY';
				else source = '';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case CST.FETCH_PRICE:
				type = 'FETCH';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public async sleep(ms: number) {
		return new Promise(resolve => {
			setTimeout(resolve, ms);
		});
	}
}

const util = new Util();
export default util;
