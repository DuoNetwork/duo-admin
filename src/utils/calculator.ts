import {Constants, IPriceFix, ITrade} from '@finbook/duo-market-data'
import * as CST from '../common/constants';
import dbUtil from './dbUtil';
import util from './util';

class Calculateor {
	public getTradesForInterval(
		trades: ITrade[],
		currentTimestamp: number,
		interval: number
	): ITrade[] {
		const oneInterval = 5 * 60 * 1000;
		const upperTime: number = currentTimestamp - interval * oneInterval;
		const lowerTime: number = upperTime - oneInterval;

		return trades.filter(
			trade => Number(trade.timestamp) <= upperTime && Number(trade.timestamp) > lowerTime
		);
	}

	public getVolumeMedianPrice(trades: ITrade[], timestamp: number): IPriceFix {
		trades.sort((a, b) => Number(a.price) - Number(b.price));
		const cumVols: number[] = [];
		let cumVol = 0;
		trades.forEach((trade, index) => {
			cumVol += Number(trade.amount);
			cumVols[index] = cumVol;
		});
		const halfTotalVol: number = cumVol / 2;
		const medianIndex = cumVols.findIndex(v => v >= halfTotalVol);
		return {
			price: Number(trades[medianIndex].price),
			volume: cumVol,
			timestamp,
			source: '',
			base: trades[0].base,
			quote: trades[0].quote
		};
	}

	public getExchangePriceFix(trades: ITrade[], currentTimestamp: number): IPriceFix {
		for (let i = 0; i < 12; i++) {
			const subTrades = this.getTradesForInterval(trades, currentTimestamp, i);
			if (subTrades.length > 0) return this.getVolumeMedianPrice(subTrades, currentTimestamp);
		}

		return {
			price: 0,
			volume: 0,
			timestamp: currentTimestamp,
			source: '',
			base: '',
			quote: ''
		};
	}

	public getWeights(rawVolume: number[]): number[] {
		const totalVol = rawVolume.reduce((a, b) => a + b, 0);
		const weightArray = rawVolume.map(v => v / totalVol);
		return weightArray;
	}

	public validateWeights(weights: number[]): boolean {
		const numOfValidExchanges = weights.length;
		for (let i = 0; i < weights.length; i++)
			if (weights[i] > CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges][i]) return false;

		return true;
	}

	public modifyWeights(weights: number[]): number[] {
		const numOfValidExchanges = weights.length;
		let isValid: boolean = this.validateWeights(weights);
		while (!isValid) {
			let sumOfCapped: number = 0;
			let sumOfUnCapped: number = 0;
			const weightCaps = CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges];
			for (let i = 0; i < weights.length; i++)
				if (weights[i] >= weightCaps[i]) {
					weights[i] = weightCaps[i];
					sumOfCapped += weightCaps[i];
				} else sumOfUnCapped += weights[i];

			for (let i = 0; i < weights.length; i++)
				if (weights[i] < weightCaps[i])
					weights[i] = (weights[i] / sumOfUnCapped) * (1 - sumOfCapped);

			isValid = this.validateWeights(weights);
		}
		return weights;
	}

	public consolidatePriceFix(exchangePriceVolume: IPriceFix[]): number {
		const filterredExchanges = exchangePriceVolume.filter(item => item.volume > 0);

		// sort based on volume from large to small
		filterredExchanges.sort((a, b) => b.volume - a.volume);
		const volumeList = filterredExchanges.map(item => item.volume);

		if (filterredExchanges.length === 0)
			// use previous priceFix
			return 0;
		else if (filterredExchanges.length === 1)
			// let finalArray: number[] = [];
			return filterredExchanges[0].price;
		else {
			util.logInfo('there are ' + filterredExchanges.length + ' valid exchanges');
			const totalVol = volumeList.reduce((a, b) => a + b, 0);
			const weights = volumeList.map(v => v / totalVol);
			const finalWeights: number[] = this.modifyWeights(weights);
			return finalWeights.reduce(
				(accum, w, index) => accum + filterredExchanges[index].price * w,
				0
			);
		}
	}

	public async getPriceFix(quote: string, base: string): Promise<IPriceFix> {
		const currentTimestamp: number = util.getUTCNowTimestamp();
		const trades = await dbUtil.readSourceData(currentTimestamp, base, quote);
		const EXCHANGES_TRADES: { [key: string]: ITrade[] } = {
			[Constants.API_BITFINEX]: [],
			[Constants.API_GEMINI]: [],
			[Constants.API_GDAX]: [],
			[Constants.API_KRAKEN]: [],
			[Constants.API_BITSTAMP]: []
		};

		trades.forEach(item => {
			if (item.source === Constants.API_BITFINEX)
				EXCHANGES_TRADES[Constants.API_BITFINEX].push(item);
			else if (item.source === Constants.API_GEMINI)
				EXCHANGES_TRADES[Constants.API_GEMINI].push(item);
			else if (item.source === Constants.API_GDAX)
				EXCHANGES_TRADES[Constants.API_GDAX].push(item);
			else if (item.source === Constants.API_KRAKEN)
				EXCHANGES_TRADES[Constants.API_KRAKEN].push(item);
			else if (item.source === Constants.API_BITSTAMP)
				EXCHANGES_TRADES[Constants.API_BITSTAMP].push(item);
		});

		const exchangePriceVolume = CST.API_LIST.map(src =>
			this.getExchangePriceFix(EXCHANGES_TRADES[src], currentTimestamp)
		);

		const priceFix: number = this.consolidatePriceFix(exchangePriceVolume);
		util.logInfo('priceFix calculated is ' + priceFix);

		if (priceFix === 0) {
			util.logInfo('no priceFix found, use the last ETH price');
			const lastPriceObj = await dbUtil.readLastPrice(base, quote);
			util.logInfo(
				'the priceFix is: ' + lastPriceObj.price + ' at timestamp ' + lastPriceObj.timestamp
			);
			return lastPriceObj;
		} else {
			const priceObj = {
				price: priceFix,
				volume: exchangePriceVolume.reduce((sum, p) => sum + p.volume, 0),
				timestamp: currentTimestamp,
				source: '',
				base: base,
				quote: quote
			};
			util.logInfo(
				'valid exchange priceFix found: ' +
					priceObj.price +
					' at ' +
					priceObj.timestamp +
					' for volume of ' +
					priceObj.volume
			);
			// save price into DB
			await dbUtil.insertPrice(priceObj);

			return priceObj;
		}
	}
}
const calculator = new Calculateor();
export default calculator;
