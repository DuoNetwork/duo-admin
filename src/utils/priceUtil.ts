// import moment from 'moment';
import DualClassWrapper from '../../../duo-contract-wrapper/src/DualClassWrapper';
import MagiWrapper from '../../../duo-contract-wrapper/src/MagiWrapper';
import apis from '../apis';
import * as CST from '../common/constants';
import { IContractPrice, IDualClassStates, IOption, IPrice } from '../common/types';
import calculator from './calculator';
import dynamoUtil from './dynamoUtil';
import util from './util';
const schedule = require('node-schedule');

class PriceUtil {
	public async startCommitPrices(magiWrapper: MagiWrapper, option: IOption) {
		const startTime = util.getUTCNowTimestamp();
		const endTime = startTime.valueOf() + 3500000;
		const commitStart = endTime + 50000;
		const rule = new schedule.RecurrenceRule();
		rule.minute = 0;

		const isStarted = await magiWrapper.isStarted();

		if (!isStarted)
			// contract is in inception state; start contract first and then commit price
			schedule.scheduleJob({ start: startTime, end: endTime, rule }, async () =>
				this.startMagi(magiWrapper, option)
			);

		schedule.scheduleJob({ start: !isStarted ? commitStart : startTime, rule }, async () =>
			this.commitPrice(magiWrapper, option)
		);
	}

	public async startMagi(magiWrapper: MagiWrapper, option: IOption) {
		const currentPrice = await calculator.getPriceFix(option.base, option.quote);
		const gasPrice = (await magiWrapper.web3Wrapper.getGasPrice()) || option.gasPrice;
		util.logInfo('gasPrice price ' + gasPrice + ' gasLimit is ' + option.gasLimit);
		return magiWrapper.startMagi(
			'',
			currentPrice.price,
			Math.floor(currentPrice.timestamp / 1000),
			{
				gasPrice: gasPrice,
				gasLimit: option.gasLimit + 50000
			}
		);
	}

	public async commitPrice(magiWrapper: MagiWrapper, option: IOption) {
		const currentPrice = await calculator.getPriceFix(option.base, option.quote);
		const gasPrice = (await magiWrapper.web3Wrapper.getGasPrice()) || option.gasPrice;
		util.logInfo('gasPrice price ' + gasPrice + ' gasLimit is ' + option.gasLimit);
		return magiWrapper.commitPrice(
			'',
			currentPrice.price,
			Math.floor(currentPrice.timestamp / 1000),
			{
				gasPrice: gasPrice,
				gasLimit: option.gasLimit
			}
		);
	}

	public async fetchPrice(
		dualClassWrappers: DualClassWrapper[],
		magiWrapper: MagiWrapper,
		option: IOption
	) {
		const isStarted = await magiWrapper.isStarted();
		if (!isStarted) {
			util.logDebug('Magi not ready, please start Magi first');
			return;
		}
		let nonce = await dualClassWrappers[0].web3Wrapper.getTransactionCount(magiWrapper.address);
		global.setInterval(async () => {
			// first checking Magi current time is set correctly
			const lastPrice: IContractPrice = await magiWrapper.getLastPrice();
			const promiseList: Array<Promise<void>> = [];
			const wrappersToCall: DualClassWrapper[] = [];

			for (const bw of dualClassWrappers) {
				const btvStates: IDualClassStates = await bw.getStates();
				if (
					btvStates.state === CST.CTD_TRADING &&
					lastPrice.timestamp - btvStates.lastPriceTime > 3000000
				)
					wrappersToCall.push(bw);
			}

			if (wrappersToCall.length) {
				const gasPrice = (await magiWrapper.web3Wrapper.getGasPrice()) || option.gasPrice;
				for (const bw of wrappersToCall) {
					promiseList.push(
						bw.fetchPrice('', {
							gasPrice: gasPrice,
							gasLimit: option.gasLimit,
							nonce: nonce
						})
					);
					nonce++;
				}

				await Promise.all(promiseList);
			}
		}, 15000);
	}

	public getBasePeriod(period: number) {
		if (period === 1) return 0;
		if (period === 60) return 1;

		throw new Error('invalid period');
	}

	public sortPricesByPairPeriod(
		prices: IPrice[],
		period: number
	): { [pair: string]: { [timestamp: number]: IPrice[] } } {
		const pairPrices: { [pair: string]: { [timestamp: number]: IPrice[] } } = {};
		prices.forEach(price => {
			const pair = price.quote + '|' + price.base;
			const timestamp = Math.floor(price.timestamp / 60000 / period);
			if (!pairPrices[pair]) pairPrices[pair] = {};
			if (!pairPrices[pair][timestamp]) pairPrices[pair][timestamp] = [];
			pairPrices[pair][timestamp].push(price);
		});
		return pairPrices;
	}

	public getPeriodPrice(prices: IPrice[], period: number): IPrice {
		prices.sort((a, b) => a.timestamp - b.timestamp);
		const first = prices[0];
		const last = prices[prices.length - 1];
		prices.sort((a, b) => -a.high + b.high);
		const high = prices[0];
		prices.sort((a, b) => a.low - b.low);
		const low = prices[0];
		return {
			source: first.source,
			open: first.open,
			high: high.high,
			low: low.low,
			close: last.close,
			volume: prices.reduce((sum, p) => sum + p.volume, 0),
			timestamp: Math.floor(first.timestamp / 60000 / period) * 60000 * period,
			period: period,
			quote: first.quote,
			base: first.base
		};
	}

	public async aggregatePrice(period: number) {
		util.logDebug(`priceUtil.aggregatePrice(${period})`);
		const now = util.getUTCNowTimestamp();
		const start = util.getPeriodStartTimestamp(now, period);
		for (const src in apis) {
			util.logInfo('------------------');
			util.logInfo(
				`[${src}]: To aggreate period:${String(
					period
				)} prices from timestamp ${util.timestampToString(start)}`
			);
			util.logInfo(
				`[${src}]: querying basePeriod:${this.getBasePeriod(
					period
				)} prices from timestamp ${util.timestampToString(start)}`
			);
			const basePrices = await dynamoUtil.getPrices(
				src,
				this.getBasePeriod(period),
				start,
				now
			);
			util.logInfo(`fetched ${basePrices.length} basePeriod prices`);
			const pairPeriodPrices = this.sortPricesByPairPeriod(basePrices, period);
			const pairPrices: { [pair: string]: IPrice[] } = {};
			for (const pair in pairPeriodPrices)
				for (const pd in pairPeriodPrices[pair]) {
					const periodPrices = pairPeriodPrices[pair][pd];
					if (periodPrices.length) {
						if (!pairPrices[pair]) pairPrices[pair] = [];
						pairPrices[pair].push(this.getPeriodPrice(periodPrices, period));
					}
				}

			util.logInfo('finished process, updating database');
			for (const pair in pairPrices)
				for (const price of pairPrices[pair]) await dynamoUtil.addPrice(price);
		}

		util.logInfo('all source processed');
	}

	public startAggregate(period: number) {
		// console.log({ period: period + '' });
		dynamoUtil.insertHeartbeat({ period: { N: period + '' } });

		global.setInterval(
			() => dynamoUtil.insertHeartbeat({ period: { N: period + '' } }),
			CST.STATUS_INTERVAL * 1000
		);
		global.setInterval(() => this.aggregatePrice(period), 30000);
	}
}

const priceUtil = new PriceUtil();

export default priceUtil;
