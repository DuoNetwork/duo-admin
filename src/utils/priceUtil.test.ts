import calculator from './calculator';
import dynamoUtil from './dynamoUtil';
import priceUtil from './priceUtil';
import util from './util';

jest.mock('node-schedule', () =>
	jest.fn().mockImplementation(() => ({
		RecurrenceRule: jest.fn(),
		scheduleJob: jest.fn()
	}))
);

test('getBasePeriod', () => {
	expect(priceUtil.getBasePeriod(1)).toBe(0);
	expect(priceUtil.getBasePeriod(60)).toBe(1);
	expect(() => priceUtil.getBasePeriod(2)).toThrowErrorMatchingSnapshot();
});

const prices = [
	{
		source: 'bitfinex',
		base: 'USD',
		quote: 'ETH',
		timestamp: 1535958681000,
		period: 0,
		open: 202,
		high: 210,
		low: 200,
		close: 205,
		volume: 0.0007
	},
	{
		source: 'kraken',
		base: 'USD',
		quote: 'ETH',
		timestamp: 1535958681000,
		period: 0,
		open: 207,
		high: 213,
		low: 200,
		close: 210,
		volume: 17.6757
	},
	{
		source: 'gemini',
		base: 'USD',
		quote: 'ETH',
		timestamp: 1535958609000,
		period: 0,
		open: 214,
		high: 220,
		low: 210,
		close: 215,
		volume: 0.0013
	}
];

const sortedPrices1 = priceUtil.sortPricesByPairPeriod(prices, 1);
const sortedPrices2 = priceUtil.sortPricesByPairPeriod(prices, 60);
test('sortPricesByPairPeriod 1', () => expect(sortedPrices1).toMatchSnapshot());
test('sortPricesByPairPeriod 60', () => expect(sortedPrices2).toMatchSnapshot());

test('getPeriodPrice 1', () => {
	for (const pair in sortedPrices1)
		for (const period in sortedPrices1[pair])
			expect(priceUtil.getPeriodPrice(sortedPrices1[pair][period], 1)).toMatchSnapshot();
});

test('getPeriodPrice 60', () => {
	for (const pair in sortedPrices2)
		for (const period in sortedPrices2[pair])
			expect(priceUtil.getPeriodPrice(sortedPrices2[pair][period], 60)).toMatchSnapshot();
});

test('aggregatePrice', async () => {
	dynamoUtil.getPrices = jest.fn((src: string) =>
		Promise.resolve(prices.filter(price => price.source === src))
	);
	dynamoUtil.addPrice = jest.fn();
	util.getUTCNowTimestamp = jest.fn(() => Math.max(...prices.map(price => price.timestamp)) + 1);

	await priceUtil.aggregatePrice(1);
	expect((dynamoUtil.addPrice as jest.Mock<Promise<boolean>>).mock.calls).toMatchSnapshot();
});

const magiWrapper = {
	web3Wrapper: {
		getTransactionCount: jest.fn(() => 100),
		getGasPrice: jest.fn(() => Promise.resolve(1000000000))
	},
	isStarted: jest.fn(() => true),
	address: 'magiAddress',
	getLastPrice: jest.fn(() => ({
		price: 100,
		timestamp: 2234567890000
	})),
	commitPrice: jest.fn(),
	startMagi: jest.fn()
} as any;
const dualClassWrapper = {
	web3Wrapper: {
		getTransactionCount: jest.fn(() => 100),
		getGasPrice: jest.fn(() => Promise.resolve(1000000000))
	},
	getStates: jest.fn(() =>
		Promise.resolve({
			state: 'Trading',
			lastPriceTime: 1234567890000
		} as any)
	),
	fetchPrice: jest.fn()
} as any;

test('fetchPrice, not started', async () => {
	global.setInterval = jest.fn();
	magiWrapper.isStarted = jest.fn(() => false);
	await priceUtil.fetchPrice('account', [dualClassWrapper], magiWrapper, 1000000000);
	expect(dualClassWrapper.web3Wrapper.getTransactionCount as jest.Mock).not.toBeCalled();
});

test('fetchPrice gasPrice', async () => {
	global.setInterval = jest.fn();
	magiWrapper.isStarted = jest.fn(() => true);

	await priceUtil.fetchPrice('account', [dualClassWrapper], magiWrapper, 1000000000);
	await (global.setInterval as jest.Mock).mock.calls[0][0]();
	expect((dualClassWrapper.fetchPrice as jest.Mock).mock.calls).toMatchSnapshot();
});

test('fetchPrice', async () => {
	global.setInterval = jest.fn();
	magiWrapper.isStarted = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	await priceUtil.fetchPrice('account', [dualClassWrapper], magiWrapper);
	await (global.setInterval as jest.Mock).mock.calls[0][0]();
	expect((dualClassWrapper.fetchPrice as jest.Mock).mock.calls).toMatchSnapshot();
});

jest.mock('node-schedule', () => ({
	RecurrenceRule: jest.fn(() => ({
		minute: 0
	})),
	scheduleJob: jest.fn()
}));

test('startCommitPrices', async () => {
	util.getUTCNowTimestamp = jest.fn(() => ({
		valueOf: jest.fn(() => 1234567890000)
	}));
	await priceUtil.startCommitPrices('account', magiWrapper, { gasPrice: 1000000000 } as any);
});

test('startCommitPrices', async () => {
	util.getUTCNowTimestamp = jest.fn(() => ({
		valueOf: jest.fn(() => 1234567890000)
	}));
	magiWrapper.isStarted = jest.fn(() => false);
	await priceUtil.startCommitPrices('account', magiWrapper, { gasPrice: 1000000000 } as any);
});

test('commitPrice', async () => {
	calculator.getPriceFix = jest.fn(() =>
		Promise.resolve({
			price: 100,
			timestamp: 1234567890000
		})
	);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => 1000000000);

	await priceUtil.commitPrice('account', magiWrapper, 'base|quote');

	expect((magiWrapper.commitPrice as jest.Mock).mock.calls).toMatchSnapshot();
});

test('commitPrice gasPrice', async () => {
	calculator.getPriceFix = jest.fn(() =>
		Promise.resolve({
			price: 100,
			timestamp: 1234567890000
		})
	);

	await priceUtil.commitPrice('account', magiWrapper, 'base|quote', 1000000000);

	expect((magiWrapper.commitPrice as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startMagi', async () => {
	calculator.getPriceFix = jest.fn(() =>
		Promise.resolve({
			price: 100,
			timestamp: 1234567890000
		})
	);

	await priceUtil.startMagi('account', magiWrapper, 'base|quote');

	expect((magiWrapper.startMagi as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startMagi gasPrice', async () => {
	calculator.getPriceFix = jest.fn(() =>
		Promise.resolve({
			price: 100,
			timestamp: 1234567890000
		})
	);

	await priceUtil.startMagi('account', magiWrapper, 'base|quote', 1000000000);

	expect((magiWrapper.startMagi as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startAggregate', async () => {
	global.setInterval = jest.fn();
	dynamoUtil.insertHeartbeat = jest.fn();
	priceUtil.aggregatePrice = jest.fn();
	await priceUtil.startAggregate(10);

	(global.setInterval as jest.Mock).mock.calls[0][0]();
	(global.setInterval as jest.Mock).mock.calls[1][0]();
	expect((dynamoUtil.insertHeartbeat as jest.Mock).mock.calls).toMatchSnapshot();
	expect((priceUtil.aggregatePrice as jest.Mock).mock.calls).toMatchSnapshot();
});
