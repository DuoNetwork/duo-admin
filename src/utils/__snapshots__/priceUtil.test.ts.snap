// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aggregatePrice 1`] = `
Array [
  Array [
    Object {
      "base": "USD",
      "close": 215,
      "high": 220,
      "low": 210,
      "open": 214,
      "period": 1,
      "quote": "ETH",
      "source": "gemini",
      "timestamp": 1535958600000,
      "volume": 0.0013,
    },
  ],
  Array [
    Object {
      "base": "USD",
      "close": 210,
      "high": 213,
      "low": 200,
      "open": 207,
      "period": 1,
      "quote": "ETH",
      "source": "kraken",
      "timestamp": 1535958660000,
      "volume": 17.6757,
    },
  ],
]
`;

exports[`commitPrice 1`] = `
Array [
  Array [
    "account",
    100,
    1234567890,
    Object {
      "gasLimit": 60000,
      "gasPrice": 1000000000,
    },
  ],
]
`;

exports[`commitPrice gasPrice 1`] = `
Array [
  Array [
    "account",
    100,
    1234567890,
    Object {
      "gasLimit": 60000,
      "gasPrice": 1000000000,
    },
  ],
  Array [
    "account",
    100,
    1234567890,
    Object {
      "gasLimit": 60000,
      "gasPrice": 1000000000,
    },
  ],
]
`;

exports[`fetchPrice 1`] = `
Array [
  Array [
    "account",
    Object {
      "gasLimit": 150000,
      "gasPrice": 1000000000,
      "nonce": 100,
    },
  ],
]
`;

exports[`fetchPrice gasPrice 1`] = `Array []`;

exports[`getBasePeriod 1`] = `"invalid period"`;

exports[`getPeriodPrice 1 1`] = `
Object {
  "base": "USD",
  "close": 215,
  "high": 220,
  "low": 210,
  "open": 214,
  "period": 1,
  "quote": "ETH",
  "source": "gemini",
  "timestamp": 1535958600000,
  "volume": 0.0013,
}
`;

exports[`getPeriodPrice 1 2`] = `
Object {
  "base": "USD",
  "close": 210,
  "high": 213,
  "low": 200,
  "open": 202,
  "period": 1,
  "quote": "ETH",
  "source": "bitfinex",
  "timestamp": 1535958660000,
  "volume": 17.676399999999997,
}
`;

exports[`getPeriodPrice 60 1`] = `
Object {
  "base": "USD",
  "close": 210,
  "high": 220,
  "low": 200,
  "open": 214,
  "period": 60,
  "quote": "ETH",
  "source": "gemini",
  "timestamp": 1535958000000,
  "volume": 17.677699999999998,
}
`;

exports[`sortPricesByPairPeriod 1 1`] = `
Object {
  "ETH|USD": Object {
    "25599310": Array [
      Object {
        "base": "USD",
        "close": 215,
        "high": 220,
        "low": 210,
        "open": 214,
        "period": 0,
        "quote": "ETH",
        "source": "gemini",
        "timestamp": 1535958609000,
        "volume": 0.0013,
      },
    ],
    "25599311": Array [
      Object {
        "base": "USD",
        "close": 205,
        "high": 210,
        "low": 200,
        "open": 202,
        "period": 0,
        "quote": "ETH",
        "source": "bitfinex",
        "timestamp": 1535958681000,
        "volume": 0.0007,
      },
      Object {
        "base": "USD",
        "close": 210,
        "high": 213,
        "low": 200,
        "open": 207,
        "period": 0,
        "quote": "ETH",
        "source": "kraken",
        "timestamp": 1535958681000,
        "volume": 17.6757,
      },
    ],
  },
}
`;

exports[`sortPricesByPairPeriod 60 1`] = `
Object {
  "ETH|USD": Object {
    "426655": Array [
      Object {
        "base": "USD",
        "close": 205,
        "high": 210,
        "low": 200,
        "open": 202,
        "period": 0,
        "quote": "ETH",
        "source": "bitfinex",
        "timestamp": 1535958681000,
        "volume": 0.0007,
      },
      Object {
        "base": "USD",
        "close": 210,
        "high": 213,
        "low": 200,
        "open": 207,
        "period": 0,
        "quote": "ETH",
        "source": "kraken",
        "timestamp": 1535958681000,
        "volume": 17.6757,
      },
      Object {
        "base": "USD",
        "close": 215,
        "high": 220,
        "low": 210,
        "open": 214,
        "period": 0,
        "quote": "ETH",
        "source": "gemini",
        "timestamp": 1535958609000,
        "volume": 0.0013,
      },
    ],
  },
}
`;

exports[`startAggregate 1`] = `
Array [
  Array [
    Object {
      "period": Object {
        "N": "10",
      },
    },
  ],
  Array [
    Object {
      "period": Object {
        "N": "10",
      },
    },
  ],
]
`;

exports[`startAggregate 2`] = `
Array [
  Array [
    10,
  ],
  Array [
    10,
  ],
]
`;

exports[`startMagi 1`] = `
Array [
  Array [
    "account",
    100,
    1234567890,
    Object {
      "gasLimit": 1000000,
      "gasPrice": 1000000000,
    },
  ],
]
`;

exports[`startMagi gasPrice 1`] = `
Array [
  Array [
    "account",
    100,
    1234567890,
    Object {
      "gasLimit": 1000000,
      "gasPrice": 1000000000,
    },
  ],
  Array [
    "account",
    100,
    1234567890,
    Object {
      "gasLimit": 1000000,
      "gasPrice": 1000000000,
    },
  ],
]
`;
